src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -439,6 +456,243 @@ instance Function Address where
+    let name' =
+          ASCIIString
+            . fmap (chr . fromIntegral)
+            . toList
+            . fromBuiltin
+            $ name
+    ASCIIString name'' <- shrink name'
+    pure
+      . TokenName
+      . toBuiltin @ByteString
+      . fromList
+      . fmap (fromIntegral . ord)
+      $ name''
+
+-- | @since 2.1.3
+deriving via BuiltinByteString instance CoArbitrary TokenName
+
+-- | @since 2.3.1
+instance Function TokenName where
+  {-# INLINEABLE function #-}
+  function = functionMap (coerce @_ @BuiltinByteString) coerce
+
+{- | This generates a phase-1 valid 'Value' meant to represent an /amount/,
+ rather than a /delta/. Thus, all of the following will hold for all generated
+ values and shrinks:
+
+ - Entries whose 'CurrencySymbol' and 'TokenName' match are unique.
+ - \'Outer\' map entries are sorted by 'CurrencySymbol'.
+ - \'Inner\' map entries are sorted by 'TokenName'.
+ - An ADA entry always exists.
+ - All entries have positive amounts.
+
+ While it is possible to have a normalized 'Value' with a zero ADA entry, no
+ such entries are currently generated.
+
+ 'Value's are shrunk only structurally: this means that we can \'drop\'
+ entries, but not shrink existing ones in either their \'inner map\' keys or
+ \'inner map\' values.
+
+ @since 2.1.3
+-}
+instance Arbitrary Value where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    Value . AssocMap.fromList <$> do
+      (Sorted cs) <- arbitrary
+      case cs of
+        [] -> do
+          adaEntry <- mkEntry ""
+          pure [adaEntry]
+        (sym : _) ->
+          if sym == ""
+            then traverse mkEntry cs
+            else do
+              adaEntry <- mkEntry ""
+              entries <- traverse mkEntry cs
+              pure $ adaEntry : entries
+    where
+      mkEntry ::
+        CurrencySymbol ->
+        Gen (CurrencySymbol, AssocMap.Map TokenName Integer)
+      mkEntry sym = do
+        (SortedNonEmpty tns) <- arbitrary
+        kvs :: [(TokenName, Positive Integer)] <-
+          traverse (\tn -> (tn,) <$> arbitrary) tns
+        pure (sym, AssocMap.fromList . coerce $ kvs)
+  {-# INLINEABLE shrink #-}
+  shrink (Value inner) =
+    Value <$> do
+      let inner' = fmap PlutusTx.toList <$> PlutusTx.toList inner
+      case inner' of
+        [] -> error "Shrinker for Value: impossible situation: given empty Value to shrink"
+        (x : xs) -> do
+          -- Shrink the 'inner map' of ADA entry structurally
+          -- We forbid empty shrinks here
+          x' :: (CurrencySymbol, [(TokenName, Integer)]) <- traverse shrinkSubseq x
+          -- Shrink the rest structurally, both outer and inner
+          -- For the outer shrink, we can allow empty prefixes
+          xs' :: [(CurrencySymbol, [(TokenName, Integer)])] <-
+            init . subsequences $ xs >>= \entry -> traverse shrinkSubseq entry
+          -- Glue these together, then rewrap
+          pure . PlutusTx.fromList . fmap (fmap PlutusTx.fromList) $ x' : xs'
```

Comment:
I don't think it will always contain `Ada` entree.

Reply:
It will. The ADA entry is first, and we only shrink what remains.